src/encoder/encoder.h

```diff
@@ -17,23 +18,56 @@
 #ifndef ENCODER_H
 #define ENCODER_H
 
+#include "util/memory.h"
 #include "util/types.h"
+#include "preferences/usersettings.h"
+#include "encoder/encodersettings.h"
+#include "encoder/encodercallback.h"
 
 class Encoder {
   public:
-    Encoder();
-    virtual ~Encoder();
+        class Format {
+            public:
+            Format(QString labelIn, QString nameIn, bool losslessIn) :
+                label(labelIn), internalName(nameIn), lossless(losslessIn) {}
+            QString label;
+            QString internalName;
+            bool lossless;
+        };
 
-    virtual int initEncoder(int bitrate, int samplerate) = 0;
+    Encoder() {}
+    virtual ~Encoder() {}
+
+    virtual int initEncoder(int samplerate, QString errorMessage) = 0;
     // encodes the provided buffer of audio.
     virtual void encodeBuffer(const CSAMPLE *samples, const int size) = 0;
-    // Adds metadata to the encoded auio, i.e., the ID3 tag. Currently only used
+    // Adds metadata to the encoded audio, i.e., the ID3 tag. Currently only used
     // by EngineRecord, EngineBroadcast does something different.
-    virtual void updateMetaData(char* artist, char* title, char* album) = 0;
+    virtual void updateMetaData(const char* artist, const char* title, const char* album) = 0;
     // called at the end when encoding is finished
     virtual void flush() = 0;
-    /**converts an OGG quality measure from 1..10 to a bitrate **/
-    static int convertToBitrate(int quality);
+    // Setup the encoder with the specific settings
+    virtual void setEncoderSettings(const EncoderSettings& settings) = 0;
+};
+
+typedef std::shared_ptr<Encoder> EncoderPointer;
```

Comment:
A shared pointer should be used when we have two or more owner of an object. 
The shared pointer counts the references and it deleted the managed object automatically if there is no reference left. 

Since c++11 we have now std::unique_ptr for situations where we have only a single owner like we have here. The uniqueness is guaranteed by removing the copy operators. 
To pass it around, you have to use std::move(), which invalidates the pointer on the caller side. 

The benefit of this approach is that it has no overhead compared to new / delete, and it documents clearly who has the ownership of an object.